#                                    DATA IMPORT
                                    



In [82]:
import os
import pandas as pd
d=pd.read_csv('Social_Network_Ads.csv',encoding='utf-8')

In [83]:
d.isnull().sum()

Age                0
EstimatedSalary    0
Purchased          0
dtype: int64

In [84]:
d

Age  EstimatedSalary  Purchased
0     19            19000          0
1     35            20000          0
2     26            43000          0
3     27            57000          0
4     19            76000          0
..   ...              ...        ...
395   46            41000          1
396   51            23000          1
397   50            20000          1
398   36            33000          0
399   49            36000          1

[400 rows x 3 columns]

# SPLIT THE DATA


In [85]:
from sklearn.model_selection import train_test_split
x=d.iloc[:,:-1]

In [86]:
y=d.iloc[:,-1]

In [87]:
x,y

(     Age  EstimatedSalary
 0     19            19000
 1     35            20000
 2     26            43000
 3     27            57000
 4     19            76000
 ..   ...              ...
 395   46            41000
 396   51            23000
 397   50            20000
 398   36            33000
 399   49            36000
 
 [400 rows x 2 columns],
 0      0
 1      0
 2      0
 3      0
 4      0
       ..
 395    1
 396    1
 397    1
 398    0
 399    1
 Name: Purchased, Length: 400, dtype: int64)

In [88]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.25,random_state=42)

x_train,x_test,y_train,y_test

# FEATURE SCALING

In [89]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
x_train=sc.fit_transform(x_train)

In [92]:
x_train

array([[ 1.8925893 ,  1.52189404],
       [ 0.1250379 ,  0.03213212],
       [ 0.9106163 , -1.31157471],
       [-1.34792161, -1.48684082],
       [-0.169554  , -0.58129926],
       [-0.56234321,  2.33980255],
       [ 1.0088136 , -1.19473064],
       [-0.75873781,  1.08372877],
       [ 2.1871812 , -1.04867555],
       [ 0.0268406 , -0.25997806],
       [-0.46414591, -1.1363086 ],
       [ 0.1250379 ,  0.03213212],
       [ 1.6961947 , -0.90262046],
       [ 1.1070109 , -0.90262046],
       [ 0.5178271 ,  1.22978386],
       [-1.05332971, -1.4576298 ],
       [-1.15152701, -1.54526286],
       [-0.0713567 ,  0.67477452],
       [ 0.4196298 , -0.46445519],
       [-0.2677513 , -0.25997806],
       [-0.85693511,  0.14897619],
       [ 0.0268406 ,  0.29503128],
       [ 0.7142217 , -1.28236369],
       [ 1.5979974 ,  1.11293979],
       [ 0.812419  , -1.36999675],
       [-1.44611891, -1.22394166],
       [-0.0713567 ,  0.14897619],
       [ 0.4196298 , -0.14313399],
       [-0.2677513 ,

In [10]:
x_test=sc.fit_transform(x_test)

In [11]:
x_test

array([[ 0.75872215, -1.41078892],
       [ 1.9109871 ,  0.5590031 ],
       [-0.83672162, -0.75419158],
       [ 0.93599368,  0.79776576],
       [-0.74808586, -1.23171691],
       [-0.65945009, -0.21697557],
       [ 0.84735791,  1.12606443],
       [-0.74808586,  0.40977643],
       [ 0.22690756,  0.17101376],
       [ 0.40417909, -0.12743957],
       [-0.21627127, -0.12743957],
       [ 1.37917251, -1.05264491],
       [-1.27990045, -0.63481024],
       [-1.54580774, -1.38094358],
       [-0.65945009,  0.52915776],
       [-0.21627127,  1.15590977],
       [ 1.29053674, -0.93326358],
       [ 0.75872215,  0.14116843],
       [ 0.13827179, -0.81388224],
       [ 1.6450798 , -0.27666624],
       [-1.36853621, -1.26156225],
       [-0.74808586,  0.32024043],
       [ 0.84735791, -1.38094358],
       [ 1.9109871 ,  0.20085909],
       [-1.63444351, -1.50032492],
       [ 1.20190098, -1.38094358],
       [ 0.40417909,  0.32024043],
       [-0.03899974, -0.48558358],
       [ 1.55644404,

In [12]:
y_test

209    0
280    1
33     0
210    1
93     0
      ..
314    0
373    1
380    0
239    1
75     1
Name: Purchased, Length: 100, dtype: int64

#                       Logistic Regression

In [51]:
from sklearn.linear_model import LogisticRegression
lgc=LogisticRegression(random_state=0)
lgc.fit(x_train,y_train)

LogisticRegression(random_state=0)

In [52]:
print(lgc)

LogisticRegression(random_state=0)


In [53]:
lgc.predict(sc.transform([[46,22000]]))

C:\Users\geeth\anaconda3\lib\site-packages\sklearn\base.py:420: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


array([0], dtype=int64)

In [54]:
import numpy as np
y_pred=lgc.predict(x_test)
#print(np.concatenate((y_pred.reshape(len(y_pred),1),y_test.reshape(len(y_test),1)),1))

In [55]:
from sklearn.metrics import confusion_matrix,accuracy_score
cm=confusion_matrix(y_pred,y_test)

In [56]:
cm
acc=accuracy_score(y_pred,y_test,normalize=False)

In [57]:
print("Logistic Regression Accuracy:",acc )

Logistic Regression Accuracy: 86


#                                          KNN NEIGHBOR
     

In [59]:
from sklearn.neighbors import KNeighborsClassifier
knn=KNeighborsClassifier(n_neighbors=5,metric='minkowski',p=2)
knn.fit(x_train,y_train)

KNeighborsClassifier()

In [60]:
y_pred=knn.predict(x_test)

In [61]:
y_pred

array([1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0,
       1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1,
       1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1], dtype=int64)

In [62]:
from sklearn.metrics import confusion_matrix,accuracy_score
cm=confusion_matrix(y_pred,y_test)
accuracy_score(y_pred,y_test,normalize=False)

92

In [64]:
print("KNN Accuracy:", accuracy_score(y_pred,y_test,normalize=False))

KNN Accuracy: 92


In [23]:
cm

array([[58,  3],
       [ 5, 34]], dtype=int64)

#                             Support VEctor Model    

In [31]:
from sklearn.svm import SVC
svc=SVC( kernel='linear',random_state=None)
print(svc.fit(x_train,y_train))

SVC(kernel='linear')


In [32]:
y_pred=svc.predict(x_test)
y_pred

array([0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1,
       0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0], dtype=int64)

In [33]:
accuracy_score(y_pred,y_test,normalize=False)

86

### SVM WITH RBF KERNEL

In [35]:
svm_rbf = SVC(kernel='rbf', random_state=0)


In [28]:
svm_rbf.fit(x_train, y_train)
y_pred_rbf = svm_rbf.predict(x_test)


In [29]:
accuracy_rbf = accuracy_score(y_pred_rbf, y_test)


In [30]:
print("SVM with RBF Kernel Accuracy:", accuracy_rbf)

SVM with RBF Kernel Accuracy: 0.95


# Decision Tree classifier

In [36]:
from sklearn.tree import DecisionTreeClassifier

In [39]:
dtc=DecisionTreeClassifier(random_state=0)

In [40]:
dtc.fit(x_train,y_train)

DecisionTreeClassifier(random_state=0)

In [41]:
y_pred_dtc=dtc.predict(x_test)

In [42]:
accuracy_dtc = accuracy_score(y_pred_dtc, y_test)


In [43]:
print("Decision Tree Accuracy:", accuracy_dtc)

Decision Tree Accuracy: 0.9


# Random Forest Classifier

In [44]:
from sklearn.ensemble import RandomForestClassifier

In [45]:
rfc = RandomForestClassifier(n_estimators=100, random_state=0)

In [46]:
rfc.fit(x_train, y_train)

RandomForestClassifier(random_state=0)

In [47]:
y_pred_rfc = rfc.predict(x_test)

In [48]:
accuracy_rfc = accuracy_score(y_pred_rfc, y_test)

In [49]:
print("Random Forest Accuracy:", accuracy_rfc)

Random Forest Accuracy: 0.92


# Gradient Boost Classifier

In [67]:
from sklearn.ensemble import GradientBoostingClassifier
gbc = GradientBoostingClassifier(n_estimators=100, random_state=0)
gbc.fit(x_train, y_train)


GradientBoostingClassifier(random_state=0)

In [68]:
y_pred_gbc = gbc.predict(x_test)
accuracy_gbc = accuracy_score(y_pred_gbc, y_test)

In [69]:
print("Gradient Boosting Accuracy:", accuracy_gbc)

Gradient Boosting Accuracy: 0.89


# Naive Bayes CLassifier

In [70]:
from sklearn.naive_bayes import GaussianNB

In [71]:
nb = GaussianNB()

In [72]:
nb.fit(x_train, y_train)

GaussianNB()

In [73]:
y_pred_nb = nb.predict(x_test)

In [74]:
accuracy_nb = accuracy_score(y_pred_nb, y_test)

In [75]:
print("Naive Bayes Accuracy:", accuracy_nb)

Naive Bayes Accuracy: 0.94


# MLP classifier

In [76]:
from sklearn.neural_network import MLPClassifier

In [77]:
mlp = MLPClassifier(hidden_layer_sizes=(100,), max_iter=1000, random_state=0)

In [78]:
mlp.fit(x_train, y_train)

MLPClassifier(max_iter=1000, random_state=0)

In [79]:
y_pred_mlp = mlp.predict(x_test)

In [80]:
accuracy_mlp = accuracy_score(y_pred_mlp, y_test)

In [81]:
print("MLP Classifier Accuracy:", accuracy_mlp)

MLP Classifier Accuracy: 0.94


Overall, SVM with RBF kernel demonstrated the best performance in this experiment scoring accuracy of 95%. Its ability to handle non-linearities well in the data contributed to this success. Following closely behind, naive bayes classifiers and MLP with 94% for its ability to analyse complex patterns, while Naive Bayes benefits from its assumption of strong independence between features. Additionally, both KNN and Random Forest classifiers exhibited respectable performance, with accuracies ranging from 90% to 92%. This demonstrates their effectiveness in handling small datasets and diverse types of data.